ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/HistoricalBlockHashProcessor.java

```diff
@@ -0,0 +1,127 @@
+
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.evm.account.MutableAccount;
+import org.hyperledger.besu.evm.operation.BlockHashOperation;
+import org.hyperledger.besu.evm.worldstate.WorldUpdater;
+import org.hyperledger.besu.plugin.data.ProcessableBlockHeader;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.tuweni.units.bigints.UInt256;
+
+/**
+ * Processes and stores historical block hashes in accordance with EIP-2935. This class is
+ * responsible for managing the storage of block hashes to support EIP-2935, which introduces
+ * historical block hash access in smart contracts.
+ */
+public class HistoricalBlockHashProcessor {
+
+  public static final Address HISTORY_STORAGE_ADDRESS = BlockHashOperation.HISTORY_STORAGE_ADDRESS;
+
+  private static final long HISTORY_SERVE_WINDOW = 8192;
+
+  private final long forkTimestamp;
+  private final long historySaveWindow;
+
+  /**
+   * Constructs a HistoricalBlockHashProcessor with a specified fork timestamp.
+   *
+   * @param forkTimestamp The timestamp at which the fork becomes active.
+   */
+  public HistoricalBlockHashProcessor(final long forkTimestamp) {
+    this(forkTimestamp, HISTORY_SERVE_WINDOW);
+  }
+
+  /**
+   * Constructs a HistoricalBlockHashProcessor with a specified fork timestamp and history save
+   * window. This constructor is primarily used for testing.
+   *
+   * @param forkTimestamp The timestamp at which the fork becomes active.
+   * @param historySaveWindow The number of blocks for which history should be saved.
+   */
+  @VisibleForTesting
+  public HistoricalBlockHashProcessor(final long forkTimestamp, final long historySaveWindow) {
+    this.forkTimestamp = forkTimestamp;
+    this.historySaveWindow = historySaveWindow;
+  }
+
+  /**
+   * Stores the historical block hashes based on the current block header. This method calculates
+   * the appropriate storage slot for the parent block hash and any additional ancestors within the
+   * history save window, then stores those hashes in the world state.
+   *
+   * @param blockchain The blockchain from which block headers are retrieved.
+   * @param worldUpdater The world state updater to store the block hashes.
+   * @param currentBlockHeader The current block header being processed.
+   */
+  public void storeHistoricalBlockHashes(
+      final Blockchain blockchain,
+      final WorldUpdater worldUpdater,
+      final ProcessableBlockHeader currentBlockHeader) {
+
+    final MutableAccount account = worldUpdater.getOrCreate(HISTORY_STORAGE_ADDRESS);
+
+    if (currentBlockHeader.getNumber() > 0) {
+      storeParentHash(account, currentBlockHeader);
+
+      BlockHeader ancestor =
+          blockchain.getBlockHeader(currentBlockHeader.getParentHash()).orElseThrow();
```

Comment:
normally it's not possible to try import a block without having the parent it will be a critical issue. maybe we need to test to be sure the behavior is correct in this case and be sure we are considering the block as problematic to allow a retry ?